### ISSUE TITLE
Share price manipulation attack vector

### ISSUE THREAD

**Issue by jennazenk** [melonproject/protocol#430]
_tldr; a fund manager was able to trade with tokens offered in an investment request, before that request was actually executed and shares were created and assigned to the investor._

The delay between an _investment request_ and the _execution_ of that request might allow for: 
(i) share price manipulation (eg. in the context of a competition) 
(ii) fund manager can trade with the tokens offered for investment before the shares are actually created and assigned to the investor. This has the following implication: if an investor `requestInvesment` and then wants to `cancelRequest`, it is possible that he might not be able to cancel his request if the fund manager already traded with the tokens offered in the _investment request_. 

Reproduction steps (or see fund named Volkan Kara with address on Kovan 0x44741eaca12736bc9adcf92feed25ad46b4c1fd8) : 

- Manager creates a fund, requests investment for 1 MLN and executes his investment request. He gets 1 share and ```GAV = 1MLN, totalSupply =1```
- Investor A makes an investment request of 10 MLN (does not execute that request)
At this stage, ```GAV = 11 MLN``` but `totalSupply = 1` (no new shares were created). The offered MLN are accounted for in the AUM although there is no token transfer in `requestInvestment` (see https://github.com/melonproject/smart-contracts/blob/master/src/Fund.sol#L186-L208).
- Fund manager places an order or takes an order using those 10 MLN.
- Investor A tries to cancel his request: not possible since the MLN were potentially traded. 
- Other option: Investor A tries to execute his request. It is unclear at this point if the execution request will go through. I personally was not able to execute the said request it just runs out of gas no matter how high I set the gas parameters. 

_Implication_ : 
- This fund now has a share price of 11 (!).
- Potential loss of fund for Investor A who can't retrieve his tokens by canceling request nor can he execute his request.
- It is an attack vector for share price manipulation in the context of competitions.

_Assumption_ : Investor A might have tried to execute his request, which didn't fully go through ; no shares were created but still the tokens were somehow available to fund manager for trading.
Any idea how this could happen @travs @GauthamN ?

**Comment by jennazenk** [comment]
Just follow up on investigation: 

The tokens need to be approved before calling `requestInvestment`. Then, even if `executeRequest` is not called, the fund contract could still call `transferFrom` on the token contract and specify an `to`address different than the fund address. 

I thought this could be the reason but actually it's not since `makeOrder` requires that the fund contract approves the tokens to the exchange:
https://github.com/melonproject/smart-contracts/blob/master/src/Fund.sol#L360

**Comment by jennazenk** [comment]
Second follow up: 

Running safeguard we found a token transfer of 10 MLN (bypassing the requestInvestment function). This transfer could not be seen on Etherscan because there are 2 different Transfer event (at least on the deployed version) and Etherscan only knows about one (see https://github.com/melonproject/smart-contracts/issues/418)

**Comment by jennazenk** [comment]
Keeping that issue open because of the second fund that was created today and that shows unexpected behavior (breaks performCalculations, most likely/precisely calcValuePerShare and returns 0 for share price ). Will document more as soon as we have more info on this one

**Comment by jennazenk** [comment]
Fund B is another fund showing suspicious behavior -may or may not be related (was created today):

According to etherscan, this is the flow of transactions of this fund:
-Made a fund
-Invested 1 MLN and executed request
- Invested 10 MLN and executed request
- Redeem(10 share) and executed request
- redeemAllOwnedAssets(1 share)
- requestInvestment(100 shares for 10 MLN) 

This flow seems to have broken the performCalculations function (maybe more precisely calcValuePerShare). Now this fund shows a share price of 0 (expected: 1), gav = nav = 0 but owns 28 MLN tokens (?)

**Comment by jennazenk** [comment]
About to close that issue since we figured out what happened. 

**Fund A**:  Fund manager was sending tokens directly to the fund (bypassing invest) by querying the faucet with his fund address. Etherscan doesn't catch the `Transfer` events emitted by the faucet transactions (see also https://github.com/melonproject/smart-contracts/issues/418) 

**Fund B**: Was also sending tokens into the fund, and events were not caught by Etherscan. However, performCalculations now breaks on this fund - edge case that was not accounted for in sharePrice calc: 
````
sharePrice = nav > 0 ? calcValuePerShare(nav, totalSupplyAccountingForFees) : toSmallestShareUnit(1); // Handle potential division through zero by defining a default value
````
_Proposed solution_ : check for `totalSupply=0` instead of `nav>0`

**Comment by travs** [comment]
This was fixed afterward
